packages/core/src/api/apis/implementations/lib/subjects.ts

```diff
@@ -0,0 +1,202 @@
+/*
+ * Copyright 2020 Spotify AB
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { Observable } from '../../../types';
+import ObservableImpl from 'zen-observable';
+
+// TODO(Rugvip): These are stopgap and probably incomplete implementations of subjects.
+// If we add a more complete Observables library they should be replaced.
+
+/**
+ * A basic implementation of ReactiveX publish subjects.
+ *
+ * A subject is a convenient way to create an observable when you want
+ * to fan out a single value to all subscribers.
+ *
+ * See http://reactivex.io/documentation/subject.html
+ */
+export class PublishSubject<T>
+  implements Observable<T>, ZenObservable.SubscriptionObserver<T> {
+  private isClosed = false;
+  private terminatingError?: Error;
+
+  private readonly observable = new ObservableImpl<T>((subscriber) => {
+    if (this.isClosed) {
+      if (this.terminatingError) {
+        subscriber.error(this.terminatingError);
+      } else {
+        subscriber.complete();
+      }
+      return () => {};
+    }
+
+    this.subscribers.add(subscriber);
+    return () => {
+      this.subscribers.delete(subscriber);
+    };
+  });
+
+  private readonly subscribers = new Set<
+    ZenObservable.SubscriptionObserver<T>
+  >();
+
+  get closed() {
+    return this.isClosed;
+  }
+
+  next(value: T) {
+    if (this.isClosed) {
+      throw new Error('PublishSubject is closed');
```

Comment:
is it really part of the contract to throw an error to the caller on all closed conditions? just double checking

Reply:
Yep, this would be if the caller for example first calls `.error()` and then `.next()`. It's a programming error by the caller and RxJs works the same way.